[AIRFLOW-100][AIRFLOW-301] Add execution_date_fn to ExternalTaskSensor and fix unit test

jlowin opened the pull request:
Dear Airflow Maintainers,

Please accept this PR that addresses the following issues:
- https://issues.apache.org/jira/browse/AIRFLOW-100
- Adds an optional callable to `ExternalTaskSensor` that makes it easy to specify the queried `execution_date`
- https://issues.apache.org/jira/browse/AIRFLOW-301
- Fixes a unit test that always fails on the last day of the month

Testing Done:
- Added new unit tests

jlowin committed:
[AIRFLOW-100] Add execution_date_fn to ExternalTaskSensor

Currently, ExternalTaskSensor only supports querying execution_dates
that are either the same as the ExternalTaskSensor's execution_date
or a fixed interval from that date (using `execution_delta`). This
adds the ability to provide a fn (`execution_date_fn`) that accepts
the current execution_date and can return any desired date for
querying. This is much more flexible. For example, it could
supply the last date of the previous month.

criccomini commented on airflow/operators/sensors.py:
```diff
@@ -179,8 +179,14 @@ class ExternalTaskSensor(BaseSensorOperator):
     :param execution_delta: time difference with the previous execution to
         look at, the default is the same execution_date as the current task.
-        For yesterday, use [positive!] datetime.timedelta(days=1)
+        For yesterday, use [positive!] datetime.timedelta(days=1) Either
```
Nit: missed a period :)

jlowin committed:
[AIRFLOW-301] Fix broken unit test

This unit tests always fails on the last day of the month, since it
tries to access a nonexistent day (like June 31st).

jlowin committed:
Add license to migration file
